packages/models/src/huggingface/text-generation.ts

```diff
@@ -0,0 +1,210 @@
+
+function headers(accessToken?: string) {
+  const headers: Record<string, string> = {
+    accept: 'application/json',
+    'content-type': 'application/json',
+  };
+  if (typeof accessToken === 'string') {
+    headers.authorization = `Bearer ${accessToken}`;
+  }
+  return headers;
+}
+
+export namespace HfChatTypes {
+  // https://huggingface.co/docs/api-inference/detailed_parameters#text-generation-task
+  export type Request = {
+    model: string;
+    stream?: boolean;
+    inputs: string;
+    parameters?: {
+      top_k?: number;
+      top_p?: number;
+      temperature?: number;
+      repetition_penalty?: number;
+      max_new_tokens?: number;
+      // In seconds
+      max_time?: number;
+      return_full_text?: boolean;
+      num_return_sequences?: number;
+      do_sample?: boolean;
+    };
+    options?: {
+      use_cache?: boolean;
+      wait_for_model?: boolean;
+    };
+  };
+
+  export type RequestOptions = {
+    accessToken?: string;
+    apiUrl?: string;
+    fetch?: typeof fetch;
+  };
+
+  export type GeneratedText = {
+    generated_text: string;
+  };
+
+  // https://huggingface.co/docs/api-inference/detailed_parameters#text-generation-task
+  export type Response = GeneratedText | GeneratedText[];
+
+  // Best documentation I could find: https://huggingface.co/docs/huggingface_hub/main/en/package_reference/inference_client#huggingface_hub.inference._text_generation.TextGenerationStreamResponse
+  // I would like to find more formal documentation of their streaming API if we can.
+  export type Chunk = {
+    token: {
+      id: number;
+      text: string;
+      logprob: number;
+      special: boolean;
+    };
+    generated_text: string;
+    // Observed this but cannot find documentation
+    details: null;
+  };
+}
+
+async function run(
+  request: HfChatTypes.Request,
+  options: HfChatTypes.RequestOptions,
+): Promise<HfChatTypes.Response> {
+  const url = options.apiUrl || HF_MODEL_API_URL + request.model;
+
+  const headers_ = headers(options.accessToken);
+  const body = JSON.stringify({ ...request, stream: false });
+  const response = await POST(url, {
+    headers: headers_,
+    body,
+    fetch: options.fetch,
+  });
+
+  return response.json();
+}
+
+async function streamBytes(
+  request: HfChatTypes.Request,
+  options: HfChatTypes.RequestOptions,
+): Promise<ReadableStream<Uint8Array>> {
+  const url = options.apiUrl || HF_MODEL_API_URL + request.model;
+
+  const headers_ = headers(options.accessToken);
+  const body = JSON.stringify({ ...request, stream: true });
+  try {
+    const response = await POST(url, {
+      headers: headers_,
+      body,
+      fetch: options.fetch,
+    });
+
+    if (!response.body) {
+      throw new HttpError('Expected response body to be a ReadableStream', response);
+    }
+
+    return response.body;
+  } catch (e) {
+    if (e instanceof HttpError) {
+      const body = await e.response.json();
```

Comment:
This will fail if the response returns non-json or 500s with non-json / no body (or combo thereof). In that case, this code would throw an error other than the original HTTP error or the error raised below. May want to wrap this in another try/catch as annoying as that may be. Guessing this isn't common, but something to consider.

Reply:
I was curious about the handling here. There might be a better way, but the current `HttpError` was never able to be decoded by the end user to see the actual error.

Happy to do a second wrap, but I'm wondering if we're missing a cleaner way.

Reply:
>but the current HttpError was never able to be decoded by the end user to see the actual error.

They are able to, it's just that node does not print the values. It looks like:

```ts
try {
  OpenAIChat.run(args, opts);
} catch (err) {
  // response is the response object from `fetch`, and thus you can call anything on it you would if you had used fetch directly
  const response = err.response;
  const body = await response.json();
}
```